File: ma_sh/Method/value.py
import torch


def toValues(
    params: torch.Tensor,
    base_values: torch.Tensor,
    phi_idxs: torch.Tensor,
) -> torch.Tensor:
    repeat_params = params[phi_idxs]

    values_matrix = repeat_params * base_values.transpose(1, 0)

    values = torch.sum(values_matrix, dim=1)

    return values

File: ma_sh/Cpp/src/value.cpp
#include "value.h"
#include <torch/script.h>

const torch::Tensor toValues(const torch::Tensor &params,
                             const torch::Tensor &base_values,
                             const torch::Tensor &phi_idxs) {
  const torch::Tensor repeat_params = params.index({phi_idxs});

  const torch::Tensor values_matrix =
      repeat_params * base_values.transpose(1, 0);

  const torch::Tensor values = torch::sum(values_matrix, 1);

  return values;
}
